{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Tips: Please install Bio before running by:__\n",
    "pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# python3 genome_simulation.py --ini='bacteria.ini' --data='Escherichia_coli_DH10B.fasta'\n",
    "from configparser import ConfigParser\n",
    "import optparse\n",
    "import random, os, sys\n",
    "import pandas as pd\n",
    "import logging\n",
    "from datetime import datetime\n",
    "import numpy as np\n",
    "import calendar;\n",
    "import time;\n",
    "from Bio import SeqIO \n",
    "from decimal import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 0. Please configure these parameters accordingly before running:\n",
    "* ALL parameters in configuration will be kept in the format of 'conf_XXXXX' for easily recognization.\n",
    "* #### For [dendrogram]:\n",
    "    * linkage_method: 'single', 'weighted', 'average', 'centroid', 'median', 'ward'\n",
    "    * metric: pdist: 'euclidean', 'minkowski', 'cityblock'(manhattan), 'seuclidean', 'cosine', 'correlation', 'hamming', 'canberra', 'jaccard', 'chebyshev', 'canberra', 'braycurtis', 'yule', 'matching' etc.\n",
    "REF: https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html;\n",
    "https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html#scipy.spatial.distance.pdist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#[lamba]\n",
    "# conf_lamba_min/max: It is on a per generation basis ... update mutation rate to different values within the range. \n",
    "# conf_lamba_min/max:Start by one generation the min rate then the next the max rate then the min rate etc\n",
    "# conf_lamba_min=0.1 \n",
    "# conf_lamba_max=0.5\n",
    "\n",
    "# conf_lamba: can have multiple lambas\n",
    "conf_lamba=0.2,0.1,0.3\n",
    "conf_generation=7\n",
    "\n",
    "#[mutation]\n",
    "# run_times: How many mutated seqs will be generated.\n",
    "conf_run_times=4 \n",
    "conf_numrows=1\n",
    "\n",
    "#[fasta]\n",
    "# conf_filepath: The path of fasta file.\n",
    "conf_filepath = '../Escherichia_coli_DH10B.fasta'\n",
    "# conf_top_num: How many bases could be kept from fasta file.\n",
    "conf_top_num = 40 \n",
    "# conf_generated_file_format: specify the format will be generated with: '.nex','.txt' etc.\n",
    "conf_generated_file_format = '.txt'\n",
    "\n",
    "# [pdist]\n",
    "if_add_hamming = 0 # 1: Add hamming distance as the column for pdist & dendrogram; 0: Remove hamming distance for dendrogram;\n",
    "\n",
    "# For VARYING LAMBDA:\n",
    "#1  alternating min-max-min-max-min-max\n",
    "#2. Random within range\n",
    "#3. Constant lambda (done)\n",
    "\n",
    "#[dendrogram]\n",
    "linkage_method = 'median'         # 'single', 'weighted', 'average', 'centroid', 'median', 'ward'\n",
    "pdistance_method =  'hamming' #  'euclidean', 'minkowski', 'cityblock'(manhattan), 'seuclidean', 'correlation', 'hamming', 'canberra', 'jaccard', 'chebyshev', 'canberra', 'braycurtis', 'yule', 'matching' etc.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 1. Function definition:\n",
    "#### A. parse_file(filename, formatname, length): Return the sequencing from file with specified format and length.\n",
    "#### B. mutate_prob(target_bases_list, prob): Mutate the base by picking from bases list with probability.\n",
    "#### C. mutation_simulate(ori_seq): Mutate the original sequencing by probability/flipping and return the mutated sequencing, and how many bases have been mutated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_file(filename, formatname, length): \n",
    "    '''\n",
    "        filename: file path\n",
    "        formatname: 'fasta', 'abi', 'genbank or gb' etc...\n",
    "        length: top number for snipping\n",
    "    '''\n",
    "    with open(filename, \"rU\") as handle:\n",
    "        for record in SeqIO.parse(handle, formatname) :\n",
    "            if len(record.seq)>length:\n",
    "                base_top = record.seq[:length]\n",
    "        return record, base_top"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Hamming distance of 2 string:\n",
    "def hamming_dist(str1, str2):\n",
    "    return sum(str1 != str2 for str1, str2 in zip(str1, str2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def mutate_prob(target_bases_list, prob1): \n",
    "    picked_base = np.random.choice(target_bases_list, 1, prob1) #Pick 1 from target list based on prob1.\n",
    "    return picked_base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def mutation_simulate(ori_seq):\n",
    "    count=0\n",
    "    mutated_seq=[]\n",
    "\n",
    "    for ori_base in ori_seq:\n",
    "        flip = np.random.uniform(0, 1)\n",
    "        #print('flip on ', ori_base, flip)\n",
    "        \n",
    "        lambda_list = list(conf_lamba)\n",
    "        picked_lamba = np.random.choice(lambda_list, 1)\n",
    "    \n",
    "        # For C base: transit to A or G by flip probability (if flip < lamda picked randomly from configuration lambda list.)\n",
    "        if (ori_base == 'C') and (flip <= picked_lamba): #C-> G or A\n",
    "            mutated_base = mutate_prob(['A','G'], flip)  #Pick 1 from target list based on flip.\n",
    "            count+=1\n",
    "            if mutated_base:\n",
    "                #print ('(Mutated', ori_base, ' to: ', mutated_base, ' flip: ', flip, 'lambda: ', picked_lamba, ')')\n",
    "                mutated_seq.append(mutated_base)   \n",
    "            else:\n",
    "                mutated_seq.append(ori_base)\n",
    "                \n",
    "        # For G base: transit to C or T by flip probability (if flip < lamda picked randomly from configuration lambda list.)       \n",
    "        elif (ori_base == 'G') and (flip <= picked_lamba): #G-> C or T\n",
    "            mutated_base = mutate_prob(['C','T'], flip)  #Pick 1 from target list based on flip.\n",
    "            count+=1\n",
    "            if mutated_base:\n",
    "                #print ('(Mutated', ori_base, ' to: ', mutated_base, ' flip: ', flip, 'lambda: ', picked_lamba, ')')\n",
    "                mutated_seq.append(mutated_base)   \n",
    "            else:\n",
    "                mutated_seq.append(ori_base)            \n",
    "        else:\n",
    "            mutated_seq.append(ori_base)\n",
    "\n",
    "    return mutated_seq, count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 2: Get the specified length of the sequences from the specified file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# 0. FAKE seq for testing:\n",
    "# ori_seq='AGCTACTGGC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sikedong/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:7: DeprecationWarning: 'U' mode is deprecated\n",
      "  import sys\n"
     ]
    }
   ],
   "source": [
    "# Get the trimmed seq string from specified file.\n",
    "record, base_top = parse_file(conf_filepath, \"fasta\", int(conf_top_num))\n",
    "ori_seq = str(base_top)\n",
    "print(ori_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 3: Generate the mutated sequences with specified times and write these mutated sequeces to file with specified format.\n",
    "- In this case, write the mutated sequeces in .nex file.  \n",
    "- The file will be named with timestamp.  \n",
    "- e.g.: __Generated file: GBAC_<Generation\\><mutation_rate\\>Seq__  \n",
    "GBAC_7_0.2 AGCTTTTCATTCTGACTGAAACGGGCAATATGTCTCTGTG  \n",
    "GBAC_7_0.2 AGATTTTCATTCTGACTGCAACGGTCAATATGTCTCTGTG  \n",
    "GBAC_7_0.2 AGCTTTTCATTCTGACTGCAACGCGCAATATTTCTATCTC  \n",
    "GBAC_7_0.2 AGATTTTCATTCTGACTTCAACGCGAAATATGTCTCTGTG  \n",
    "GBAC_7_0.2 AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1565855122\n"
     ]
    }
   ],
   "source": [
    "# Generate file name with timestamp.\n",
    "timestamp = calendar.timegm(time.gmtime())\n",
    "print(timestamp)\n",
    "mut_seq_nexfilepath = \"file_generated\\\\\"+\"SimSeqFile_\"+ str(timestamp) + conf_generated_file_format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 4: Generate the sequences and differences for output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def compare_seq(seq1, seq2):\n",
    "    diff_seq = ''\n",
    "    diff_index = ''\n",
    "    diff=''\n",
    "    \n",
    "    if len(seq1) == len(seq2):\n",
    "        for i in range(0, len(seq1)):\n",
    "            if seq1[i]!=seq2[i]:\n",
    "                diff_seq+=seq2[i]\n",
    "                diff_index+=str(i)+','\n",
    "        diff = diff_seq +' '+ diff_index\n",
    "        \n",
    "    else:\n",
    "        return null\n",
    "    \n",
    "    return diff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Sike difference marker\n",
    "def dna_seq_compare(ori_seq, mutated_seq):\n",
    "    len1 = len(ori_seq)\n",
    "    len2 = len(mutated_seq)\n",
    "    mismatches = []\n",
    "    diff_pos =[]\n",
    "    for pos in range (0, min(len1, len2)):\n",
    "        if ori_seq[pos]!= mutated_seq[pos]:\n",
    "            mismatches.append(\"|\")\n",
    "            diff_pos.append(pos)\n",
    "        else:\n",
    "            mismatches.append(\" \")\n",
    "    mismatches = \"\".join(mismatches)\n",
    "\n",
    "    return diff_pos, mismatches+'\\n'+ mutated_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Sike compare two sequences, equal = 0, different = 1\n",
    "def dna_seq_compare_equality(seq1, seq2):\n",
    "    len1 = len(seq1)\n",
    "    len2 = len(seq2)\n",
    "    result = []\n",
    "    if len(seq1) == len(seq2):\n",
    "        for i in range(0, min(len1, len2)):\n",
    "            if seq1[i]!=seq2[i]:\n",
    "                result.append('1')\n",
    "            else:\n",
    "                result.append('0')\n",
    "    result = \"\".join(result)\n",
    "\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seq</th>\n",
       "      <th>hamming_distance</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>...</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>39</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   id                                       seq  hamming_distance  0  1  2  3  \\\n",
       "0   0  AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG                 0  0  0  0  0   \n",
       "\n",
       "   4  5  6 ...  30  31  32  33  34  35  36  37  38  39  \n",
       "0  0  0  0 ...   0   0   0   0   0   0   0   0   0   0  \n",
       "\n",
       "[1 rows x 43 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create dataframe to store all seqs(original and mutated), hamming distancr and sepcific difference.\n",
    "multi_mutated_seq =''\n",
    "diff_str=''\n",
    "multi_mutated_seq ='>Ori\\n' + ori_seq +'\\n'\n",
    "\n",
    "list_all_hamming= []\n",
    "\n",
    "for i in range(0, len(ori_seq)):\n",
    "    list_all_hamming.append(str(i))\n",
    "    \n",
    "df_mutseq_hamming = pd.DataFrame(columns=['id', 'seq', 'hamming_distance']+list_all_hamming).append({\"id\":0, \"seq\":ori_seq, \"hamming_distance\":0},ignore_index=True)\n",
    "df_mutseq_hamming.fillna(0) #Fill the NaN value with 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hamming distance: 2\n",
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG\n",
      " |     |                                \n",
      "ACCTTTTGATTCTGACTGCAACGGGCAATATGTCTCTGTG\n",
      "\n",
      "0100000100000000000000000000000000000000\n",
      "Hamming distance: 5\n",
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG\n",
      "           |     |   ||            |    \n",
      "AGCTTTTCATTGTGACTTCAAGCGGCAATATGTCTGTGTG\n",
      "\n",
      "0000000000010000010001100000000000010000\n",
      "Hamming distance: 5\n",
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG\n",
      " |     |       |  |  |                  \n",
      "ACCTTTTAATTCTGAGTGAAAAGGGCAATATGTCTCTGTG\n",
      "\n",
      "0100000100000001001001000000000000000000\n",
      "Hamming distance: 0\n",
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG\n",
      "                                        \n",
      "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG\n",
      "\n",
      "0000000000000000000000000000000000000000\n",
      "1 time(s) mutation: CG 1,7,\n",
      "2 time(s) mutation: GTGCG 11,17,21,22,35,\n",
      "3 time(s) mutation: CAGAA 1,7,15,18,21,\n",
      "4 time(s) mutation:  \n",
      "\n",
      "Result:\n",
      "0100000100000000000000000000000000000000\n",
      "0000000000010000010001100000000000010000\n",
      "0100000100000001001001000000000000000000\n",
      "0000000000000000000000000000000000000000\n",
      "\n"
     ]
    }
   ],
   "source": [
    "multi_mutated_seq=''\n",
    "diff_str=''\n",
    "mark_str = ''\n",
    "compare_ori_mut='Result:'+'\\n'\n",
    "multi_mutated_seq='Original_Sequencing: ' + ori_seq+ '\\n'\n",
    "\n",
    "for run_time in range(conf_run_times):\n",
    "    # print('\\nMutating...', run_time+1, 'times: ')\n",
    "    mutated_seq_file_list,count = mutation_simulate(ori_seq) #Tuple return: mutation_simulate(): mutated_seq, count. Get mutated_seq.\n",
    "        \n",
    "    #seq_num = 'GBAC_' + str(conf_generation)+'_'+ str(conf_lamba)\n",
    "    mutated_seq= ''.join([''.join(base) for base in mutated_seq_file_list])\n",
    "    #multi_mutated_seq += seq_num +':'+ mutated_seq+'\\n'\n",
    "    \n",
    "    hamming_distance = hamming_dist(ori_seq,mutated_seq)\n",
    "    print('Hamming distance:', hamming_distance)\n",
    "    \n",
    "    df_mutseq_hamming = df_mutseq_hamming.append({\"id\":run_time+1, \"seq\":mutated_seq, \"hamming_distance\":hamming_distance},ignore_index=True)\n",
    "    \n",
    "    diff_str += str(run_time+1) +' time(s) mutation: '+ compare_seq(ori_seq, mutated_seq)+'\\n'\n",
    "    \n",
    "    #Sike difference marker, compare two sequences, equal = 0, different = 1\n",
    "    str_diff_result = ori_seq + '\\n' + dna_seq_compare(ori_seq, mutated_seq)[1] + '\\n' \n",
    "    print(str_diff_result)\n",
    "    mark_str += str_diff_result + '\\n'\n",
    "    \n",
    "    compare_res = dna_seq_compare_equality(ori_seq, mutated_seq)\n",
    "    print(compare_res)\n",
    "    compare_ori_mut += compare_res + '\\n'\n",
    "    \n",
    "    #Ziwei write dataframe\n",
    "    list_diff_result = dna_seq_compare_equality(ori_seq, mutated_seq)[0]\n",
    "    for diff in list_diff_result:\n",
    "        df_mutseq_hamming.set_value(run_time+1, str(diff),1)\n",
    "\n",
    "print(diff_str)\n",
    "print(compare_ori_mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def write_file(filepath, content):\n",
    "    with open(filepath, \"w\") as file:\n",
    "        file.write(content)\n",
    "        print('Done. Wrote to: ', filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Wrote to:  file_generated\\SimSeqFile_1565855122.txt\n"
     ]
    }
   ],
   "source": [
    "write_file(mut_seq_nexfilepath, diff_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Wrote to:  file_generated\\Diff_Ori_Mut_1565855122.txt\n"
     ]
    }
   ],
   "source": [
    "diff_filepath = \"file_generated\\\\\" +\"Diff_Ori_Mut_\" + str(timestamp) + conf_generated_file_format\n",
    "write_file(diff_filepath, mark_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done. Wrote to:  file_generated\\Comp_Ori_Mut_1565855122.txt\n"
     ]
    }
   ],
   "source": [
    "compare_filepath = \"file_generated\\\\\" +\"Comp_Ori_Mut_\" + str(timestamp) + conf_generated_file_format\n",
    "write_file(compare_filepath, compare_ori_mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>seq</th>\n",
       "      <th>hamming_distance</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>...</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>39</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ACCTTTTCATTCTGACTGCAACCTGCAATATGTCTATGTG</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ACCTTTTCATTCTGACTCCAAAGGGCAATATTTCTGTGTG</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ATGTTTTCATTCTGACTGCAACGGCCAATATGTCTGTGTG</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ATGTTTTCATTCTGACTGAAACGGGCAATATCTCTCTGTG</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>AGCTTTTCATTCTGACTGCAACGGGAAATATTTGTCTGTC</td>\n",
       "      <td>4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>AGCTTTTCATTGTGAATGCAACCGGGAATATGTCTATGTG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>ACCTTTTAATTCTGACTGGAACTCGAAATATGTCTCTGTG</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>AGCTTTTAATTCTTAGTCCAAAGGGCAATATGTATCTGTG</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>AGCTTTTCATTCTTACTGCAACGGGCAATATGTGTCTGTG</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>AGCTTTTCATTCTGACTCCAACGGGCAATATGTCTCTGTT</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>3</td>\n",
       "      <td>AGCTTTTCATTCTGACTCCAACGGGGAATATGTCTCTGTC</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4</td>\n",
       "      <td>AGCTTTTGATTCTGACTGCAACGGGCAATATCTCTCTGTG</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>ACCTTTTGATTCTGACTGCAACGGGCAATATGTCTCTGTG</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2</td>\n",
       "      <td>AGCTTTTCATTGTGACTTCAAGCGGCAATATGTCTGTGTG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>3</td>\n",
       "      <td>ACCTTTTAATTCTGAGTGAAAAGGGCAATATGTCTCTGTG</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>4</td>\n",
       "      <td>AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    id                                       seq  hamming_distance    0    1  \\\n",
       "0    1  ACCTTTTCATTCTGACTGCAACCTGCAATATGTCTATGTG                 4  0.0  0.0   \n",
       "1    2  ACCTTTTCATTCTGACTCCAAAGGGCAATATTTCTGTGTG                 5  1.0  0.0   \n",
       "2    3  ATGTTTTCATTCTGACTGCAACGGCCAATATGTCTGTGTG                 4  1.0  0.0   \n",
       "3    4  ATGTTTTCATTCTGACTGAAACGGGCAATATCTCTCTGTG                 4  1.0  0.0   \n",
       "4    1  AGCTTTTCATTCTGACTGCAACGGGAAATATTTGTCTGTC                 4  1.0  0.0   \n",
       "5    2  AGCTTTTCATTGTGAATGCAACCGGGAATATGTCTATGTG                 5  0.0  0.0   \n",
       "6    3  ACCTTTTAATTCTGACTGGAACTCGAAATATGTCTCTGTG                 6  0.0  0.0   \n",
       "7    4  AGCTTTTAATTCTTAGTCCAAAGGGCAATATGTATCTGTG                 6  0.0  0.0   \n",
       "8    1  AGCTTTTCATTCTTACTGCAACGGGCAATATGTGTCTGTG                 2  0.0  0.0   \n",
       "9    2  AGCTTTTCATTCTGACTCCAACGGGCAATATGTCTCTGTT                 2  0.0  0.0   \n",
       "10   3  AGCTTTTCATTCTGACTCCAACGGGGAATATGTCTCTGTC                 3  0.0  0.0   \n",
       "11   4  AGCTTTTGATTCTGACTGCAACGGGCAATATCTCTCTGTG                 2  0.0  0.0   \n",
       "12   1  ACCTTTTGATTCTGACTGCAACGGGCAATATGTCTCTGTG                 2  0.0  0.0   \n",
       "13   2  AGCTTTTCATTGTGACTTCAAGCGGCAATATGTCTGTGTG                 5  0.0  0.0   \n",
       "14   3  ACCTTTTAATTCTGAGTGAAAAGGGCAATATGTCTCTGTG                 5  0.0  0.0   \n",
       "15   4  AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTG                 0  0.0  0.0   \n",
       "\n",
       "      2    3    4    5    6 ...    30   31   32   33   34   35   36   37   38  \\\n",
       "0   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "1   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "2   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "3   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "4   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "5   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "6   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "7   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "8   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "9   0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "10  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "11  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "12  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "13  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "14  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "15  0.0  0.0  0.0  0.0  0.0 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0   \n",
       "\n",
       "     39  \n",
       "0   0.0  \n",
       "1   0.0  \n",
       "2   0.0  \n",
       "3   0.0  \n",
       "4   0.0  \n",
       "5   0.0  \n",
       "6   0.0  \n",
       "7   0.0  \n",
       "8   0.0  \n",
       "9   0.0  \n",
       "10  0.0  \n",
       "11  0.0  \n",
       "12  0.0  \n",
       "13  0.0  \n",
       "14  0.0  \n",
       "15  0.0  \n",
       "\n",
       "[16 rows x 43 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate dataframe with hamming distance\n",
    "df_mutseq_hamming = df_mutseq_hamming.fillna(0)\n",
    "df_mutseq_hamming"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 5: Clustering:\n",
    "#### scipy.cluster.hierarchy.linkage():\n",
    "* The following linkage methods are used to compute the distance d(s, t) between two clusters s and t.\n",
    "* The following are methods for calculating the distance between the newly formed cluster u and each v:\n",
    "    * method=’single’ assigns\n",
    "    * method=’complete’ assigns\n",
    "    * method=’average’ assigns\n",
    "    * method=’centroid’ assigns\n",
    "    * method=’median’ assigns\n",
    "    * method=’ward’\n",
    "    \n",
    "#### Note:\n",
    "The resulting matrix Z is informing each step of the agglomerative clustering by informing the first two columns of which cluster indices were merged.\n",
    "The third column is the distance between those clusters, and the fourth column is the number of original samples contained in that newly merged cluster.\n",
    "* it is necessary to group similar genes.\n",
    "\n",
    "#### Visualizing the clusters:\n",
    "* A good way to visualise this is with a dendrogram, which shows at which inter-cluster distance each merge occurred. From there, it is possible to select a distance where clusters are clear (indicated with the horizontal black lines)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Sike - dendrogram for latest 4 sequences\n",
    "import scipy \n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from scipy.spatial import distance\n",
    "from scipy.spatial.distance import pdist\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get the sequences(all the mutated sequences are stored in the df)\n",
    "df_mutseq_hamming = df_mutseq_hamming.iloc[-4:]\n",
    "df_mutseq_hamming_lastest = df_mutseq_hamming[df_mutseq_hamming.columns[2:]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hamming_distance</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>30</th>\n",
       "      <th>31</th>\n",
       "      <th>32</th>\n",
       "      <th>33</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>37</th>\n",
       "      <th>38</th>\n",
       "      <th>39</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    hamming_distance    0    1    2    3    4    5    6    7    8 ...    30  \\\n",
       "12                 2  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0   \n",
       "13                 5  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0   \n",
       "14                 5  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0   \n",
       "15                 0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0 ...   0.0   \n",
       "\n",
       "     31   32   33   34   35   36   37   38   39  \n",
       "12  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "13  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "14  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "15  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "\n",
       "[4 rows x 41 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mutseq_hamming_lastest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get the pairwise distance matrix:\n",
    "* Pairwise distance between observations, by returning a condensed distance matrix Y.\n",
    "#### scipy.spatial.distance.pdist(X, metric='XX', *args, **kwargs):\n",
    "    * Pairwise distances between observations in n-dimensional space:\n",
    "    * #### X:\n",
    "    * An m by n array of m original observations in an n-dimensional space.\n",
    "    * #### metric:\n",
    "    * The distance metric to use. The distance function can be ‘braycurtis’, ‘canberra’, ‘chebyshev’, ‘cityblock’, ‘correlation’, ‘cosine’, ‘dice’, ‘euclidean’, ‘hamming’, ‘jaccard’, ‘jensenshannon’, ‘kulsinski’, ‘mahalanobis’, ‘matching’, ‘minkowski’, ‘rogerstanimoto’, ‘russellrao’, ‘seuclidean’, ‘sokalmichener’, ‘sokalsneath’, ‘sqeuclidean’, ‘yule’."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAEWCAYAAAB42tAoAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4wLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvpW3flQAAFxNJREFUeJzt3XuUJnV95/H3hxkuytU4o0ZAIZGL\nI3FFW1dkN6ISBVSICUchy3FBIlFEdxVd2WiM4sY9ohtP9JDoxHVVXBHERMaIOyYouGtEGfDGgCTj\neGHEyKCACIZL/O4fVZ157On59TMN1f3M8H6dM6efqvpV1fep6e5P/+rye1JVSJK0JTssdgGSpMlm\nUEiSmgwKSVKTQSFJajIoJElNBoUkqcmgkLZBSf4wyfsXuw49MBgUmnhJTkjy5SR3JLmpf316kixw\nHScnqSR/OmP+b/fzPzjmdt6c5CNbsd8jkmwYnVdVb6uq3x93G9J9YVBooiU5E/gz4B3AI4CHAy8D\nDgd2WoSSvg28KMnSkXkvBv5hEWqRFoRBoYmVZE/gbOD0qrqoqm6vzler6j9U1V19u52TvDPJ95P8\nKMl7kzyoX3ZEkg1Jzux7Iz9McsrIPra47hb8E/BN4Dn9+r8CPA1YNbLNzXoASb6b5MgkRwF/SBc2\nP0vy9X75KUmuS3J7kvVJ/qCfvyvwGeCRffufJXnkzF5JkmOTrE1ya5LLkjx2xr5fm+QbSW5LckGS\nXebxX6IHKINCk+wwYGfg4jnavR04EHgC8Bhgb+BNI8sfAezZzz8VODfJQ8ZcdzYfputFAJzQ13fX\n3G8Hqur/AG8DLqiq3arq3/SLbgKeB+wBnAK8K8kTq+oO4Gjgxr79blV14+g2kxwInA/8Z2A5cAnw\nqSSjPa4XAkcB+wOPB04ep14JDApNtmXAzVV17/SMJH/f/9X88yS/2V+neCnw6qr6SVXdTveL+ISR\n7dwDnF1V91TVJcDPgIPGXHc2fw0c0fd4XkwXHPdJVX26qr7d95guBz4L/PsxV38R8Omq+tuqugd4\nJ/Agup7OtHdX1Y1V9RPgU3TBKI1l6dxNpEXzY2BZkqXTYVFVTwPoT+3sQPcX9IOBq0aubQdYMrqd\n0bAB7gR2G3PdzVTVz5N8GngjsKyqvpjk6Hm/y+79HA38MV3vZoe+rm+Oufojge+N1PeLJDfQ9Y6m\n/dPI6zv7daSx2KPQJPsS3Smd4xptbgZ+Djyuqvbq/+1ZVbuNsf37su6HgTOB82ZZdgfdL3oAkiyh\nC6VpvzRkc5KdgU/Q9QQeXlV70Z0+ymztZ3Ej8OiR7QXYF/jBGO9DmpNBoYlVVbcCbwH+PMnxSXZL\nskOSJwC79m1+Afwl3Tn9hwEk2TvJc8bY/rzXBS4Hfgt4zyzL/gHYJclzk+xI1/PYeWT5j4D9kkz/\n/O3UL98I3Nv3Lp49o/1D+1Nds7kQeG6SZ/X7O5MuYP9+jPchzcmg0ESrqnOA1wD/he6C74+A9wGv\nZ9MvwtcD64ArkvwU+DvgoDF3Ma91+2sJl/bn/Gcuuw04HXg/3V/1dwCjd0F9vP/64yRX99dGXkX3\nC/8W4PcYuYuqqr5Fd7F6fX995pdOG1XV9cBJdKF1M/B84PlVdfcY71+aU/zgIklSiz0KSVLTYEGR\n5AP9A07XbGF5krw7ybr+QaAnDlWLJGn+huxRfJDuAZ8tORo4oP93GvAXA9YiSZqnwYKiqr4AbHah\nb8RxwIf7i4JXAHsl+dWh6pEkzc9iPnC3N3DDyPSGft4PZzZMchpdr4Ndd931SQcffPCCFChJ24ur\nrrrq5qpaPnfLzS1mUMw2RPSst2BV1UpgJcDU1FStWbNmyLokabuT5Htzt5rdYt71tIHu6dFp+9A9\nYSpJmiCLGRSrgBf3dz89FbitqjY77SRJWlyDnXpKcj5wBN2gbhvoBjzbEaCq3ks3ls0xdE/F3kk3\ntLIkacIMFhRVdeIcywt4xVD7lyTdP3wyW5LUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkU\nkqQmg0KS1GRQSJKaDApJUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJ\najIoJElNBoUkqcmgkCQ1GRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQm\ng0KS1GRQSJKaDApJUtOgQZHkqCTXJ1mX5KxZlj8qyeeTfDXJN5IcM2Q9kqStN1hQJFkCnAscDawA\nTkyyYkazNwIXVtWhwAnAnw9VjyRpfobsUTwFWFdV66vqbuBjwHEz2hSwR/96T+DGAeuRJM3DkEGx\nN3DDyPSGft6oNwMnJdkAXAK8crYNJTktyZokazZu3DhErZKkLRgyKDLLvJoxfSLwwaraBzgGOC/J\nZjVV1cqqmqqqqeXLlw9QqiRpS4YMig3AviPT+7D5qaVTgQsBqupLwC7AsgFrkiRtpSGD4krggCT7\nJ9mJ7mL1qhltvg88CyDJY+mCwnNLkjRBBguKqroXOANYDVxHd3fT2iRnJzm2b3Ym8NIkXwfOB06u\nqpmnpyRJi2jpkBuvqkvoLlKPznvTyOtrgcOHrEGSdN/4ZLYkqcmgkCQ1GRSSpCaDQpLUZFBIkpoM\nCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS1GRQSJKaDApJUpNBIUlqMigkSU0GhSSpyaCQ\nJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJajIoJElNBoUkqcmgkCQ1GRSSpCaDQpLUZFBIkpoMCklS\nk0EhSWoyKCRJTQaFJKnJoJAkNQ0aFEmOSnJ9knVJztpCmxcmuTbJ2iQfHbIeSdLWWzrUhpMsAc4F\nfgvYAFyZZFVVXTvS5gDgvwKHV9UtSR42VD2SpPkZskfxFGBdVa2vqruBjwHHzWjzUuDcqroFoKpu\nGrAeSdI8DBkUewM3jExv6OeNOhA4MMkXk1yR5KjZNpTktCRrkqzZuHHjQOVKkmYzZFBklnk1Y3op\ncABwBHAi8P4ke222UtXKqpqqqqnly5ff74VKkrZsyKDYAOw7Mr0PcOMsbS6uqnuq6jvA9XTBIUma\nEEMGxZXAAUn2T7ITcAKwakabTwLPAEiyjO5U1PoBa5IkbaWxgiLJgUkuTXJNP/34JG9srVNV9wJn\nAKuB64ALq2ptkrOTHNs3Ww38OMm1wOeB11XVj+f7ZiRJ979UzbxsMEuj5HLgdcD7qurQft41VXXI\nwPVtZmpqqtasWbPQu5WkbVqSq6pqaj7rjnvq6cFV9ZUZ8+6dzw4lSduWcYPi5iS/Tn/XUpLjgR8O\nVpUkaWKM+2T2K4CVwMFJfgB8BzhpsKokSRNjrKCoqvXAkUl2BXaoqtuHLUuSNCnGvevpbUn2qqo7\nqur2JA9J8t+GLk6StPjGvUZxdFXdOj3Rj810zDAlSZImybhBsSTJztMTSR4E7NxoL0naTox7Mfsj\nwKVJ/hfdnU8vAT40WFWSpIkx7sXsc5J8E3gW3WB/b62q1YNWJkmaCGN/cFFVfQb4zIC1SJIm0Lh3\nPf1Okn9McluSnya5PclPhy5OkrT4xu1RnAM8v6quG7IYSdLkGfeupx8ZEpL0wDRuj2JNkgvoPj/i\nrumZVfVXg1QlSZoY4wbFHsCdwLNH5hVgUEjSdm7c22NPGboQSdJkGisokuwCnAo8Dthlen5VvWSg\nuiRJE2Lci9nnAY8AngNcDuwDOIKsJD0AjBsUj6mqPwLuqKoPAc8FfmO4siRJk2LcoLin/3prkkOA\nPYH9BqlIkjRRxr3raWWShwBvBFYBuwF/NFhVkqSJMW5QXNp/BsUXgF8DSLL/YFVJkibGuKeePjHL\nvIvuz0IkSZOp2aNIcjDdLbF7JvmdkUV7MHKbrCRp+zXXqaeDgOcBewHPH5l/O/DSoYqSJE2OZlBU\n1cXAxUkOq6ovLVBNkqQJMu41ihck2SPJjkkuTXJzkpMGrUySNBHGDYpnV9VP6U5DbQAOBF43WFWS\npIkxblDs2H89Bji/qn4yUD2SpAkz7nMUn0ryLeDnwOlJlgP/PFxZkqRJMVaPoqrOAg4DpqrqHuAO\n4LghC5MkTYa5nqN4ZlV9bvQZiiSjTfzgIknazs116uk3gc/RPUNRQGZ8NSgkaTs3V1DcnuQ1wDVs\nCgj615KkB4C5gmK3/utBwJOBi+nC4vl0AwRKkrZzzYvZVfWWqnoLsAx4YlW9tqrOBJ5E9yl3TUmO\nSnJ9knVJzmq0Oz5JJZna2jcgSRrWuM9RPAq4e2T6bub44KIkS4BzgaOBFcCJSVbM0m534FXAl8es\nRZK0gMZ9juI84CtJ/pru+sQLgA/Nsc5TgHVVtR4gycfobqm9dka7twLnAK8dt2hJ0sIZ9zmKPwFO\nAW4BbgVOqar/PsdqewM3jExv6Of9qySHAvtW1d+0NpTktCRrkqzZuHHjOCVLku4n4/YoqKqrgau3\nYtuZZd6/3i2VZAfgXcDJY+x7JbASYGpqyjuuJGkBjXuNYj42APuOTO8D3DgyvTtwCHBZku8CTwVW\neUFbkibLkEFxJXBAkv2T7AScAKyaXlhVt1XVsqrar6r2A64Ajq2qNQPWJEnaSoMFRVXdC5wBrAau\nAy6sqrVJzk5y7FD7lSTdv8a+RjEfVXUJcMmMeW/aQtsjhqxFkjQ/Q556kiRtBwwKSVKTQSFJajIo\nJElNBoUkqcmgkCQ1GRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS\n1GRQSJKaDApJUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJajIoJElN\nBoUkqcmgkCQ1GRSSpCaDQpLUZFBIkpoMCklS06BBkeSoJNcnWZfkrFmWvybJtUm+keTSJI8esh5J\n0tYbLCiSLAHOBY4GVgAnJlkxo9lXgamqejxwEXDOUPVIkuZnyB7FU4B1VbW+qu4GPgYcN9qgqj5f\nVXf2k1cA+wxYjyRpHoYMir2BG0amN/TztuRU4DOzLUhyWpI1SdZs3LjxfixRkjSXIYMis8yrWRsm\nJwFTwDtmW15VK6tqqqqmli9ffj+WKEmay9IBt70B2Hdkeh/gxpmNkhwJvAF4elXdNWA9kqR5GLJH\ncSVwQJL9k+wEnACsGm2Q5FDgfcCxVXXTgLVIkuZpsKCoqnuBM4DVwHXAhVW1NsnZSY7tm70D2A34\neJKvJVm1hc1JkhbJkKeeqKpLgEtmzHvTyOsjh9y/JOm+88lsSVKTQSFJajIoJElNBoUkqcmgkCQ1\nGRSSpCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTQaFJKnJoJAkNRkUkqQmg0KS1GRQSJKaDApJUpNB\nIUlqMigkSU0GhSSpyaCQJDUZFJKkJoNCktRkUEiSmgwKSVKTQSFJajIoJElNBoUkqcmgkCQ1GRSS\npCaDQpLUZFBIkpoMCklSk0EhSWoyKCRJTYMGRZKjklyfZF2Ss2ZZvnOSC/rlX06y35D1SJK23mBB\nkWQJcC5wNLACODHJihnNTgVuqarHAO8C3j5UPZKk+RmyR/EUYF1Vra+qu4GPAcfNaHMc8KH+9UXA\ns5JkwJokSVtp6YDb3hu4YWR6A/Bvt9Smqu5NchvwUODm0UZJTgNO6yfvSnLNIBVve5Yx41g9gHks\nNvFYbOKx2OSg+a44ZFDM1jOoebShqlYCKwGSrKmqqfte3rbPY7GJx2ITj8UmHotNkqyZ77pDnnra\nAOw7Mr0PcOOW2iRZCuwJ/GTAmiRJW2nIoLgSOCDJ/kl2Ak4AVs1oswr4j/3r44HPVdVmPQpJ0uIZ\n7NRTf83hDGA1sAT4QFWtTXI2sKaqVgH/EzgvyTq6nsQJY2x65VA1b4M8Fpt4LDbxWGzisdhk3sci\n/gEvSWrxyWxJUpNBIUlqmtigcPiPTcY4Fq9Jcm2SbyS5NMmjF6POhTDXsRhpd3ySSrLd3ho5zrFI\n8sL+e2Ntko8udI0LZYyfkUcl+XySr/Y/J8csRp1DS/KBJDdt6VmzdN7dH6dvJHniWBuuqon7R3fx\n+9vArwE7AV8HVsxoczrw3v71CcAFi133Ih6LZwAP7l+//IF8LPp2uwNfAK4Apha77kX8vjgA+Crw\nkH76YYtd9yIei5XAy/vXK4DvLnbdAx2L3wSeCFyzheXHAJ+he4btqcCXx9nupPYoHP5jkzmPRVV9\nvqru7CevoHtmZXs0zvcFwFuBc4B/XsjiFtg4x+KlwLlVdQtAVd20wDUulHGORQF79K/3ZPNnurYL\nVfUF2s+iHQd8uDpXAHsl+dW5tjupQTHb8B97b6lNVd0LTA//sb0Z51iMOpXuL4bt0ZzHIsmhwL5V\n9TcLWdgiGOf74kDgwCRfTHJFkqMWrLqFNc6xeDNwUpINwCXAKxemtImztb9PgGGH8Lgv7rfhP7YD\nY7/PJCcBU8DTB61o8TSPRZId6EYhPnmhClpE43xfLKU7/XQEXS/z/yY5pKpuHbi2hTbOsTgR+GBV\n/Y8kh9E9v3VIVf1i+PImyrx+b05qj8LhPzYZ51iQ5EjgDcCxVXXXAtW20OY6FrsDhwCXJfku3TnY\nVdvpBe1xf0Yurqp7quo7wPV0wbG9GedYnApcCFBVXwJ2oRsw8IFmrN8nM01qUDj8xyZzHov+dMv7\n6EJiez0PDXMci6q6raqWVdV+VbUf3fWaY6tq3oOhTbBxfkY+SXejA0mW0Z2KWr+gVS6McY7F94Fn\nASR5LF1QbFzQKifDKuDF/d1PTwVuq6ofzrXSRJ56quGG/9jmjHks3gHsBny8v57//ao6dtGKHsiY\nx+IBYcxjsRp4dpJrgX8BXldVP168qocx5rE4E/jLJK+mO9Vy8vb4h2WS8+lONS7rr8f8MbAjQFW9\nl+76zDHAOuBO4JSxtrsdHitJ0v1oUk89SZImhEEhSWoyKCRJTQaFJKnJoJAkNRkU2iYkeXiSjyZZ\nn+SqJF9K8oKB9/nmJD9I8rUk/5jkr5KsGFn+/unpJJck2at//aok1yX53/0ox3/Xb+NFQ9YrDWUi\nn6OQRvWDPX4S+FBV/V4/79HAQjwr8q6qeme/zxcBn0vyG1W1sap+f7pRVY0OW306cHRVfad/qGnH\nqnrCuDtMsrQfv0yaCPYotC14JnB3/8AQAFX1vap6D0CSJUnekeTKfoz9P+jnH5HksiQXJflW/xd+\n+mVPSnJ53ztZPc4ImlV1AfBZYDqsLpseHiTJd5MsS/JeuuGuVyV5PfAR4Al9j+LXt7TffltvS3I5\n8J+SLE/yif49XZnk8L7dm9N95sBlfe/qVdP1JXlx//6/nuS8ft6s25G2hj0KbQseB1zdWH4q3VAE\nT06yM/DFJJ/tlx3ar38j8EXg8CRfBt4DHFdVG/uewp8ALxmjlquBg7e0sKpelm6U1mdU1c39vl5b\nVc9LsiNwXmO/e1XV0wHSfcjQu6rq/yV5FN1Tx4/t2x1MNzTH7sD1Sf6CbniONwCH9/v9lb7tnzW2\nI43FoNA2J8m5wL+j62U8GXg28Pgkx/dN9qQb/O5u4CtVtaFf72vAfsCtdIMH/m3fwVgCzDnezfTu\n70PpB82x3wtGXh8JrMimj1jZI8nu/etP9wM/3pXkJuDhdL2ui6rqZoCq+klrO1V1+314H3qAMSi0\nLVgL/O70RFW9It0gd9OD/QV4ZVWtHl0pyRHA6Ei6/0L3PR9gbVUdNo9aDh3Z79aaa793jLzeATis\nqn7+SxvofuFv6T3NNh7PrNuRtobXKLQt+BywS5KXj8x78Mjr1cDL+1M7JDkwya6N7V0PLE/3uQQk\n2THJ4+YqIsnv0vVezt/aNzCP/X4WOGNk33NdDL8UeGGSh/btp089be12pM0YFJp4/Sifvw08Pcl3\nknyF7mNwX983eT9wLXB1ug+Vfx+N3nL/cZnHA29P8nXga8DTttD81dO3xwInAc+sqnkNT72V+30V\nMNVfnL4WeNkc215Ld73j8n7bfzqf7UizcfRYSVKTPQpJUpNBIUlqMigkSU0GhSSpyaCQJDUZFJKk\nJoNCktT0/wEMOOS2XY/ZRQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10c846668>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Y=pdist(df_mutseq_hamming_lastest,'cityblock')\n",
    "Z=linkage(Y,'average')\n",
    "plt.title('Gene Mutation')\n",
    "plt.xlabel('Gene Diifference')\n",
    "plt.ylabel('distance')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
